{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Publications markdown generator for academicpages\n",
    "\n",
    "Takes a TSV of publications with metadata and converts them for use with [academicpages.github.io](academicpages.github.io). This is an interactive Jupyter notebook ([see more info here](http://jupyter-notebook-beginner-guide.readthedocs.io/en/latest/what_is_jupyter.html)). The core python code is also in `publications.py`. Run either from the `markdown_generator` folder after replacing `publications.tsv` with one containing your data.\n",
    "\n",
    "TODO: Make this work with BibTex and other databases of citations, rather than Stuart's non-standard TSV format and citation style.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data format\n",
    "\n",
    "The TSV needs to have the following columns: pub_date, title, venue, excerpt, citation, site_url, and paper_url, with a header at the top. \n",
    "\n",
    "- `excerpt` and `paper_url` can be blank, but the others must have values. \n",
    "- `pub_date` must be formatted as YYYY-MM-DD.\n",
    "- `url_slug` will be the descriptive part of the .md file and the permalink URL for the page about the paper. The .md file will be `YYYY-MM-DD-[url_slug].md` and the permalink will be `https://[yourdomain]/publications/YYYY-MM-DD-[url_slug]`\n",
    "\n",
    "This is how the raw file looks (it doesn't look pretty, use a spreadsheet or other program to edit and create)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pub_date\ttitle\tvenue\texcerpt\tcitation\turl_slug\tpaper_url\n",
      "2017-07-26\tUnsupervised Extraction of Stable Expression Signatures from Public Compendia with an Ensemble of Neural Networks.\tCell Systems\tThis paper is about eADAGE.\t\"Tan J, Doing G, Lewis KA, Price CE, Chen KM, Cady KC, Perchuk B, Laub MT, Hogan\n",
      "DA, Greene CS. Unsupervised Extraction of Stable Expression Signatures from\n",
      "Public Compendia with an Ensemble of Neural Networks. Cell Syst. 2017 Jul\n",
      "26;5(1):63-71.e6. doi: 10.1016/j.cels.2017.06.003. Epub 2017 Jul 12. PubMed\n",
      "[citation] PMID: 28711280, PMCID: PMC5532071\"\tUnsupervised-Extraction-of-Stable-Expression-Signatures-from-Public-Compendia-with-an-Ensemble-of-Neural-Networks\thttp://academicpages.github.io/files/Unsupervised Extraction of Stable Expression Signatures from Public Compendia with an Ensemble of Neural Networks.pdf\n",
      "2018-01-04\t\"Draft Genome Sequence of a Violacein-Producing\n",
      "Iodobacter sp. from the Hudson Valley Watershed\"\tGenome Announcements\tThis paper is about strain BJB302.\t\"Doing G, Perron GG, Jude BA. Draft Genome Sequence of a Violacein-Producing Iodobacter sp. from the Hudson Valley Watershed. Genome Announc. 2018 Jan 4;6(1). pii:PMCID: PMC5754501 e01428-17. doi: 10.1128/genomeA.01428-17. PubMed [citation] PMID: 29301892, \"\tDraft-Genome-Sequence-of-a-Violacein-Producing-Iodobacter-sp.-from-the-Hudson-Valley-Watershed\thttp://academicpages.github.io/files/Draft Genome Sequence of a Violacein-Producing Iodobacter sp. from the Hudson Valley Watershed.pdf\n",
      "2018-03-26\tA Multimodal Strategy Used by a Large c-di-GMP Network\tJournal of Bacteriology\tThis paper is about Pseudomonas DGC-PDE interactions.\t\"Dahlstrom KM, Collins AJ, Doing G, Taroni JN, Gauvin TJ, Greene CS, Hogan DA,\n",
      "O'Toole GA. A Multimodal Strategy Used by a Large c-di-GMP Network. J Bacteriol. \n",
      "2018 Mar 26;200(8). pii: e00703-17. doi: 10.1128/JB.00703-17. Print 2018 Apr 15. \n",
      "PubMed [citation] PMID: 29311282, PMCID: PMC5869470\"\tA-Multimodal-Strategy-Used-by-a-Large-c-di-GMP-Network\thttp://academicpages.github.io/files/A Multimodal Strategy Used by a Large c-di-GMP Network.pdf\n",
      "2018-01-18\t\"Genome Sequences of\n",
      "Phenotypically Distinct Janthinobacterium sp. Isolates Cultured from the Hudson\n",
      "Valley Watershed\"\tGenome Announcements\tThis paper is about strains from the BJB collection.\t\"Bettina AM, Doing G, O'Brien K, Perron GG, Jude BA. Draft Genome Sequences of\n",
      "Phenotypically Distinct Janthinobacterium sp. Isolates Cultured from the Hudson\n",
      "Valley Watershed. Genome Announc. 2018 Jan 18;6(3). pii: e01426-17. doi:\n",
      "10.1128/genomeA.01426-17. PubMed [citation] PMID: 29348334, PMCID: PMC5773719\"\tGenome-Sequences-of-Phenotypically-Distinct-Janthinobacterium-sp.-Isolates-Cultured-from-the-Hudson-Valley-Watershed\thttp://academicpages.github.io/files/Draft Genome Sequences of Phenotypically Distinct Janthinobacterium sp. Isolates Cultured from the Hudson Valley Watershed.pdf\n",
      "2018-07-03\t\"PathCORE-T: identifying and\n",
      "visualizing globally co-occurring pathways in large transcriptomic compendia\"\tBioData Mining\tThis paper is about gene set and pathway enrichment analysis.\t\"Chen KM, Tan J, Way GP, Doing G, Hogan DA, Greene CS. PathCORE-T: identifying and\n",
      "visualizing globally co-occurring pathways in large transcriptomic compendia.\n",
      "BioData Min. 2018 Jul 3;11:14. doi: 10.1186/s13040-018-0175-7. eCollection 2018. \n",
      "PubMed [citation] PMID: 29988723, PMCID: PMC6029133\"\tPathCORE-T:-identifying-and-visualizing-globally-co-occurring-pathways-in-large-transcriptomic-compendia\thttp://academicpages.github.io/PathCORE-T_ identifying and visualizing globally co-occurring pathways in large transcriptomic compendia.pdf\n",
      "2019-05-22\t\"Ethanol Stimulates Trehalose Production through a SpoT-DksA-AlgU-Dependent\n",
      "Pathway in Pseudomonas aeruginosa\"\tJournal of Bacteriology\tThis paper is about the P. aeruginosa response to ethanol.\t\"Harty CE, Martins D, Doing G, Mould DL, Clay ME, Occhipinti P, Nguyen D, Hogan\n",
      "DA. Ethanol Stimulates Trehalose Production through a SpoT-DksA-AlgU-Dependent\n",
      "Pathway in Pseudomonas aeruginosa. J Bacteriol. 2019 May 22;201(12). pii:\n",
      "e00794-18. doi: 10.1128/JB.00794-18. Print 2019 Jun 15. PubMed [citation] PMID:\n",
      "30936375, PMCID: PMC6531624\"\tEthanol-Stimulates-Trehalose-Production-through-a-SpoT-DksA-AlgU-Dependent-Pathway-in-Pseudomonas-aeruginosa\thttp://academicpages.github.io/Ethanol Stimulates Trehalose Production through a SpoT-DksA-AlgU-Dependent Pathway in Pseudomonas aeruginosa.pdf\n",
      "2020-08-19\t\"Conditional antagonism in\n",
      "co-cultures of Pseudomonas aeruginosa and Candida albicans: An intersection of\n",
      "ethanol and phosphate signaling distilled from dual-seq transcriptomics\"\tPLoS Genetics\tThis paper is about co-culture and dual-seq transcriptomics.\t\"Doing G, Koeppen K, Occipinti P, Harty CE, Hogan DA. Conditional antagonism in\n",
      "co-cultures of Pseudomonas aeruginosa and Candida albicans: An intersection of\n",
      "ethanol and phosphate signaling distilled from dual-seq transcriptomics. PLoS\n",
      "Genet. 2020 Aug 19;16(8):e1008783. doi: 10.1371/journal.pgen.1008783. eCollection\n",
      "2020 Aug. PubMed [citation] PMID: 32813693, PMCID: PMC7480860\"\tConditional-antagonism-in-co-cultures-of-Pseudomonas-aeruginosa-and-Candida-albicans:-An-intersection-of-ethanol-and-phosphate-signaling-distilled-from-dual-seq transcriptomics\thttp://academicpages.github.io/Conditional antagonism in co-cultures of Pseudomonas aeruginosa and Candida albicans_ An intersection of ethanol and phosphate signaling distilled from dual-seq transcriptomics.pdf\n",
      "2020-11-03\t\"Correcting for experiment-specific \n",
      "variability in expression compendia can remove underlying signals\"\tGigascience\tThis paper is about gene expression compendia.\t\"Lee AJ, Park Y, Doing G, Hogan DA, Greene CS. Correcting for experiment-specific \n",
      "variability in expression compendia can remove underlying signals. Gigascience.\n",
      "2020 Nov 3;9(11). pii: giaa117. doi: 10.1093/gigascience/giaa117. PubMed\n",
      "[citation] PMID: 33140829, PMCID: PMC7607552\"\tCorrecting-for-experiment-specific-variability-in-expression-compendia-can-remove-underlying-signals\thttp://academicpages.github.io/giaa117.pdf\n",
      "2022-06-16\t\"CF-Seq, an accessible web application for rapid \n",
      "re-analysis of cystic fibrosis pathogen RNA sequencing studies\"\tScientific Data\tThis paper is about development of an app for re-analysis of public data.\t\"Neff SL, Hampton TH, Puerner C, Cengher L, Doing G, Lee AJ, Koeppen K, Cheung AL,\n",
      "Hogan DA, Cramer RA, Stanton BA. CF-Seq, an accessible web application for rapid \n",
      "re-analysis of cystic fibrosis pathogen RNA sequencing studies. Sci Data. 2022\n",
      "Jun 16;9(1):343. doi: 10.1038/s41597-022-01431-1. PubMed [citation] PMID:\n",
      "35710652, PMCID: PMC9203545\"\t\"CF-Seq,-an-accessible-web-application-for-rapid-re-analysis-of-cystic-fibrosis-pathogen-RNA-sequencing-studies\"\t\"http://academicpages.github.io/CF-Seq, an accessible web application for rapid re-analysis of cystic fibrosis pathogen RNA sequencing studies.pdf\"\n",
      "2022-08-10\t\"Using genome-wide\n",
      "expression compendia to study microorganisms\"\tComputational and Structural Biotechnology\tThis is a review.\t\"Lee AJ, Reiter T, Doing G, Oh J, Hogan DA, Greene CS. Using genome-wide\n",
      "expression compendia to study microorganisms. Comput Struct Biotechnol J. 2022\n",
      "Aug 10;20:4315-4324. doi: 10.1016/j.csbj.2022.08.012. eCollection 2022. Review.\n",
      "PubMed [citation] PMID: 36016717, PMCID: PMC9396250\"\tUsing-genome-wide-expression-compendia-to-study-microorganisms\thttp://academicpages.github.io/Using genome-wide expression compendia to study microorganisms.pdf\n",
      "2022-10-20\t\"SOPHIE: Generative Neural Networks Separate Common and Specific\n",
      "Transcriptional Responses\"\tGenomics Proteomics Bioinformatics\tThis paper is about simulating gene expression data.\t\"Lee AJ, Mould DL, Crawford J, Hu D, Powers RK, Doing G, Costello JC, Hogan DA,\n",
      "Greene CS. SOPHIE: Generative Neural Networks Separate Common and Specific\n",
      "Transcriptional Responses. Genomics Proteomics Bioinformatics. 2022\n",
      "Oct;20(5):912-927. doi: 10.1016/j.gpb.2022.09.011. Epub 2022 Oct 7. PubMed\n",
      "[citation] PMID: 36216026, PMCID: PMC10025681\"\tSOPHIE:-Generative-Neural-Networks-Separate-Common-and-Specific-Transcriptional-Responses. \thttp://academicpages.github.io/SOPHIE_ Generative Neural Networks Separate Common and Specific Transcriptional Responses.pdf\n",
      "2023-02-23\t\"Compendium-Wide Analysis\n",
      "of Pseudomonas aeruginosa Core and Accessory Genes Reveals Transcriptional\n",
      "Patterns across Strains PAO1 and PA14\"\tmSystems\tThis paper is about PAO1 and PA14 transcriptomics.\t\"Lee AJ, Doing G, Neff SL, Reiter T, Hogan DA, Greene CS. Compendium-Wide Analysis\n",
      "of Pseudomonas aeruginosa Core and Accessory Genes Reveals Transcriptional\n",
      "Patterns across Strains PAO1 and PA14. mSystems. 2023 Feb 23;8(1):e0034222. doi: \n",
      "10.1128/msystems.00342-22. Epub 2022 Dec 21. PubMed [citation] PMID: 36541762,\n",
      "PMCID: PMC9948736\"\tCompendium-Wide-Analysis-of-Pseudomonas-aeruginosa-Core-and-Accessory-Genes-Reveals-Transcriptional-Patterns-across-Strains-PAO1-and-PA14\thttp://academicpages.github.io/Compendium-Wide Analysis of Pseudomonas aeruginosa Core and Accessory Genes Reveals Transcriptional Patterns across Strains PAO1 and PA14.pdf\n",
      "2023-02-23\t\"Computationally Efficient Assembly of Pseudomonas aeruginosa Gene Expression\n",
      "Compendia\"\tmSystems\tThis paper is about compendia of microbial gene expression.\t\"Doing G, Lee AJ, Neff SL, Reiter T, Holt JD, Stanton BA, Greene CS, Hogan DA.\n",
      "Computationally Efficient Assembly of Pseudomonas aeruginosa Gene Expression\n",
      "Compendia. mSystems. 2023 Feb 23;8(1):e0034122. doi: 10.1128/msystems.00341-22.\n",
      "Epub 2022 Dec 21. PubMed [citation] PMID: 36541761, PMCID: PMC9948711\"\tComputationally-Efficient-Assembly-of-Pseudomonas-aeruginosa-Gene-Expression-Compendia\thttp://academicpages.github.io/Computationally Efficient Assembly of Pseudomonas aeruginosa Gene Expression Compendia.pdf\n",
      "2024-04-02\tPseudomonas aeruginosa transcriptome analysis of metal restriction in ex vivo cystic fibrosis sputum\tMicrobiology Spectrum\tThis paper looks at P. aeruginosa gene expression in ex vivo sputum.\t\"Neff SL, Doing G, Reiter T, Hampton TH, Greene CS, Hogan DA. Pseudomonas aeruginosa transcriptome analysis of metal restriction in ex vivo cystic fibrosis sputum. Microbiol Spectr. 2024 Apr 2;12(4):e0315723. doi: 10.1128/spectrum.03157-23. Epub 2024 Feb 22. PMID: 38385740; PMCID: PMC10986534.\"\tPseudomonas-aeruginosa-transcriptome-analysis-of-metal-restriction-in-ex-vivo-cystic-fibrosis-sputum\thttp://academicpages.github.io/neff_and_doing-et-al-2024-pseudomonas-aeruginosa-transcriptome-analysis-of-metal-restriction-in-ex-vivo-cystic-fibrosis-sputum.pdf\n",
      "2024-10-18\t\"E. PathDash, pathway activation analysis of publicly available pathogen gene expression data\"\tmSystems\tThis is a tool for the re-analysis of public gene expression datasets.\t\"Taub L, Hampton TH, Sarkar S, Doing G, Neff SL, Finger CE, Ferreira Fukutani K, Stanton BA.2024.E.PathDash, pathway activation analysis of publicly available pathogen gene expression data. mSystems9:e01030-24.https://doi.org/10.1128/msystems.01030-24\"\tE-PathDash-pathway-activation-analysis-of-publicly-available-pathogen-gene-expression-data\thttp://academicpages.github.io/taub-et-al-2024-e-pathdash-pathway-activation-analysis-of-publicly-available-pathogen-gene-expression-data.pdf\n",
      "2024-12-01\tTEAL-seq: Targeted Expression Analysis Sequencing\tbioRxiv\tThis pre-print shares a method for targeted gene expression analysis in low-biomass samples.\t\"Doing G, Shanbhag P, Bell I, Cassidy S, Motakis E, Aiken E, Oh J, Adams MD. 2024. TEAL-seq: Targeted Expression Analysis Sequencing. bioRxiv 2024.11.26.625462.\"\tTEAL-seq-Targeted-Expression-Analysis-Sequencing\thttp://academicpages.github.io/TEAL-seq.pdf\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t\n",
      "\t\t\t\t\t\t"
     ]
    }
   ],
   "source": [
    "!cat publications.tsv\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import pandas\n",
    "\n",
    "We are using the very handy pandas library for dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import TSV\n",
    "\n",
    "Pandas makes this easy with the read_csv function. We are using a TSV, so we specify the separator as a tab, or `\\t`.\n",
    "\n",
    "I found it important to put this data in a tab-separated values format, because there are a lot of commas in this kind of data and comma-separated values can get messed up. However, you can modify the import statement, as pandas also has read_excel(), read_json(), and others."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pub_date</th>\n",
       "      <th>title</th>\n",
       "      <th>venue</th>\n",
       "      <th>excerpt</th>\n",
       "      <th>citation</th>\n",
       "      <th>url_slug</th>\n",
       "      <th>paper_url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2017-07-26</td>\n",
       "      <td>Unsupervised Extraction of Stable Expression S...</td>\n",
       "      <td>Cell Systems</td>\n",
       "      <td>This paper is about eADAGE.</td>\n",
       "      <td>Tan J, Doing G, Lewis KA, Price CE, Chen KM, C...</td>\n",
       "      <td>Unsupervised-Extraction-of-Stable-Expression-S...</td>\n",
       "      <td>http://academicpages.github.io/files/Unsupervi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2018-01-04</td>\n",
       "      <td>Draft Genome Sequence of a Violacein-Producing...</td>\n",
       "      <td>Genome Announcements</td>\n",
       "      <td>This paper is about strain BJB302.</td>\n",
       "      <td>Doing G, Perron GG, Jude BA. Draft Genome Sequ...</td>\n",
       "      <td>Draft-Genome-Sequence-of-a-Violacein-Producing...</td>\n",
       "      <td>http://academicpages.github.io/files/Draft Gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2018-03-26</td>\n",
       "      <td>A Multimodal Strategy Used by a Large c-di-GMP...</td>\n",
       "      <td>Journal of Bacteriology</td>\n",
       "      <td>This paper is about Pseudomonas DGC-PDE intera...</td>\n",
       "      <td>Dahlstrom KM, Collins AJ, Doing G, Taroni JN, ...</td>\n",
       "      <td>A-Multimodal-Strategy-Used-by-a-Large-c-di-GMP...</td>\n",
       "      <td>http://academicpages.github.io/files/A Multimo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2018-01-18</td>\n",
       "      <td>Genome Sequences of\\nPhenotypically Distinct J...</td>\n",
       "      <td>Genome Announcements</td>\n",
       "      <td>This paper is about strains from the BJB colle...</td>\n",
       "      <td>Bettina AM, Doing G, O'Brien K, Perron GG, Jud...</td>\n",
       "      <td>Genome-Sequences-of-Phenotypically-Distinct-Ja...</td>\n",
       "      <td>http://academicpages.github.io/files/Draft Gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2018-07-03</td>\n",
       "      <td>PathCORE-T: identifying and\\nvisualizing globa...</td>\n",
       "      <td>BioData Mining</td>\n",
       "      <td>This paper is about gene set and pathway enric...</td>\n",
       "      <td>Chen KM, Tan J, Way GP, Doing G, Hogan DA, Gre...</td>\n",
       "      <td>PathCORE-T:-identifying-and-visualizing-global...</td>\n",
       "      <td>http://academicpages.github.io/PathCORE-T_ ide...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2019-05-22</td>\n",
       "      <td>Ethanol Stimulates Trehalose Production throug...</td>\n",
       "      <td>Journal of Bacteriology</td>\n",
       "      <td>This paper is about the P. aeruginosa response...</td>\n",
       "      <td>Harty CE, Martins D, Doing G, Mould DL, Clay M...</td>\n",
       "      <td>Ethanol-Stimulates-Trehalose-Production-throug...</td>\n",
       "      <td>http://academicpages.github.io/Ethanol Stimula...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2020-08-19</td>\n",
       "      <td>Conditional antagonism in\\nco-cultures of Pseu...</td>\n",
       "      <td>PLoS Genetics</td>\n",
       "      <td>This paper is about co-culture and dual-seq tr...</td>\n",
       "      <td>Doing G, Koeppen K, Occipinti P, Harty CE, Hog...</td>\n",
       "      <td>Conditional-antagonism-in-co-cultures-of-Pseud...</td>\n",
       "      <td>http://academicpages.github.io/Conditional ant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2020-11-03</td>\n",
       "      <td>Correcting for experiment-specific \\nvariabili...</td>\n",
       "      <td>Gigascience</td>\n",
       "      <td>This paper is about gene expression compendia.</td>\n",
       "      <td>Lee AJ, Park Y, Doing G, Hogan DA, Greene CS. ...</td>\n",
       "      <td>Correcting-for-experiment-specific-variability...</td>\n",
       "      <td>http://academicpages.github.io/giaa117.pdf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2022-06-16</td>\n",
       "      <td>CF-Seq, an accessible web application for rapi...</td>\n",
       "      <td>Scientific Data</td>\n",
       "      <td>This paper is about development of an app for ...</td>\n",
       "      <td>Neff SL, Hampton TH, Puerner C, Cengher L, Doi...</td>\n",
       "      <td>CF-Seq,-an-accessible-web-application-for-rapi...</td>\n",
       "      <td>http://academicpages.github.io/CF-Seq, an acce...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2022-08-10</td>\n",
       "      <td>Using genome-wide\\nexpression compendia to stu...</td>\n",
       "      <td>Computational and Structural Biotechnology</td>\n",
       "      <td>This is a review.</td>\n",
       "      <td>Lee AJ, Reiter T, Doing G, Oh J, Hogan DA, Gre...</td>\n",
       "      <td>Using-genome-wide-expression-compendia-to-stud...</td>\n",
       "      <td>http://academicpages.github.io/Using genome-wi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022-10-20</td>\n",
       "      <td>SOPHIE: Generative Neural Networks Separate Co...</td>\n",
       "      <td>Genomics Proteomics Bioinformatics</td>\n",
       "      <td>This paper is about simulating gene expression...</td>\n",
       "      <td>Lee AJ, Mould DL, Crawford J, Hu D, Powers RK,...</td>\n",
       "      <td>SOPHIE:-Generative-Neural-Networks-Separate-Co...</td>\n",
       "      <td>http://academicpages.github.io/SOPHIE_ Generat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2023-02-23</td>\n",
       "      <td>Compendium-Wide Analysis\\nof Pseudomonas aerug...</td>\n",
       "      <td>mSystems</td>\n",
       "      <td>This paper is about PAO1 and PA14 transcriptom...</td>\n",
       "      <td>Lee AJ, Doing G, Neff SL, Reiter T, Hogan DA, ...</td>\n",
       "      <td>Compendium-Wide-Analysis-of-Pseudomonas-aerugi...</td>\n",
       "      <td>http://academicpages.github.io/Compendium-Wide...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2023-02-23</td>\n",
       "      <td>Computationally Efficient Assembly of Pseudomo...</td>\n",
       "      <td>mSystems</td>\n",
       "      <td>This paper is about compendia of microbial gen...</td>\n",
       "      <td>Doing G, Lee AJ, Neff SL, Reiter T, Holt JD, S...</td>\n",
       "      <td>Computationally-Efficient-Assembly-of-Pseudomo...</td>\n",
       "      <td>http://academicpages.github.io/Computationally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2024-04-02</td>\n",
       "      <td>Pseudomonas aeruginosa transcriptome analysis ...</td>\n",
       "      <td>Microbiology Spectrum</td>\n",
       "      <td>This paper looks at P. aeruginosa gene express...</td>\n",
       "      <td>Neff SL, Doing G, Reiter T, Hampton TH, Greene...</td>\n",
       "      <td>Pseudomonas-aeruginosa-transcriptome-analysis-...</td>\n",
       "      <td>http://academicpages.github.io/neff_and_doing-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2024-10-18</td>\n",
       "      <td>E. PathDash, pathway activation analysis of pu...</td>\n",
       "      <td>mSystems</td>\n",
       "      <td>This is a tool for the re-analysis of public g...</td>\n",
       "      <td>Taub L, Hampton TH, Sarkar S, Doing G, Neff SL...</td>\n",
       "      <td>E-PathDash-pathway-activation-analysis-of-publ...</td>\n",
       "      <td>http://academicpages.github.io/taub-et-al-2024...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2024-12-01</td>\n",
       "      <td>TEAL-seq: Targeted Expression Analysis Sequencing</td>\n",
       "      <td>bioRxiv</td>\n",
       "      <td>This pre-print shares a method for targeted ge...</td>\n",
       "      <td>Doing G, Shanbhag P, Bell I, Cassidy S, Motaki...</td>\n",
       "      <td>TEAL-seq-Targeted-Expression-Analysis-Sequencing</td>\n",
       "      <td>http://academicpages.github.io/TEAL-seq.pdf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      pub_date                                              title  \\\n",
       "0   2017-07-26  Unsupervised Extraction of Stable Expression S...   \n",
       "1   2018-01-04  Draft Genome Sequence of a Violacein-Producing...   \n",
       "2   2018-03-26  A Multimodal Strategy Used by a Large c-di-GMP...   \n",
       "3   2018-01-18  Genome Sequences of\\nPhenotypically Distinct J...   \n",
       "4   2018-07-03  PathCORE-T: identifying and\\nvisualizing globa...   \n",
       "5   2019-05-22  Ethanol Stimulates Trehalose Production throug...   \n",
       "6   2020-08-19  Conditional antagonism in\\nco-cultures of Pseu...   \n",
       "7   2020-11-03  Correcting for experiment-specific \\nvariabili...   \n",
       "8   2022-06-16  CF-Seq, an accessible web application for rapi...   \n",
       "9   2022-08-10  Using genome-wide\\nexpression compendia to stu...   \n",
       "10  2022-10-20  SOPHIE: Generative Neural Networks Separate Co...   \n",
       "11  2023-02-23  Compendium-Wide Analysis\\nof Pseudomonas aerug...   \n",
       "12  2023-02-23  Computationally Efficient Assembly of Pseudomo...   \n",
       "13  2024-04-02  Pseudomonas aeruginosa transcriptome analysis ...   \n",
       "14  2024-10-18  E. PathDash, pathway activation analysis of pu...   \n",
       "15  2024-12-01  TEAL-seq: Targeted Expression Analysis Sequencing   \n",
       "\n",
       "                                         venue  \\\n",
       "0                                 Cell Systems   \n",
       "1                         Genome Announcements   \n",
       "2                      Journal of Bacteriology   \n",
       "3                         Genome Announcements   \n",
       "4                               BioData Mining   \n",
       "5                      Journal of Bacteriology   \n",
       "6                                PLoS Genetics   \n",
       "7                                  Gigascience   \n",
       "8                              Scientific Data   \n",
       "9   Computational and Structural Biotechnology   \n",
       "10          Genomics Proteomics Bioinformatics   \n",
       "11                                    mSystems   \n",
       "12                                    mSystems   \n",
       "13                       Microbiology Spectrum   \n",
       "14                                    mSystems   \n",
       "15                                     bioRxiv   \n",
       "\n",
       "                                              excerpt  \\\n",
       "0                         This paper is about eADAGE.   \n",
       "1                  This paper is about strain BJB302.   \n",
       "2   This paper is about Pseudomonas DGC-PDE intera...   \n",
       "3   This paper is about strains from the BJB colle...   \n",
       "4   This paper is about gene set and pathway enric...   \n",
       "5   This paper is about the P. aeruginosa response...   \n",
       "6   This paper is about co-culture and dual-seq tr...   \n",
       "7      This paper is about gene expression compendia.   \n",
       "8   This paper is about development of an app for ...   \n",
       "9                                   This is a review.   \n",
       "10  This paper is about simulating gene expression...   \n",
       "11  This paper is about PAO1 and PA14 transcriptom...   \n",
       "12  This paper is about compendia of microbial gen...   \n",
       "13  This paper looks at P. aeruginosa gene express...   \n",
       "14  This is a tool for the re-analysis of public g...   \n",
       "15  This pre-print shares a method for targeted ge...   \n",
       "\n",
       "                                             citation  \\\n",
       "0   Tan J, Doing G, Lewis KA, Price CE, Chen KM, C...   \n",
       "1   Doing G, Perron GG, Jude BA. Draft Genome Sequ...   \n",
       "2   Dahlstrom KM, Collins AJ, Doing G, Taroni JN, ...   \n",
       "3   Bettina AM, Doing G, O'Brien K, Perron GG, Jud...   \n",
       "4   Chen KM, Tan J, Way GP, Doing G, Hogan DA, Gre...   \n",
       "5   Harty CE, Martins D, Doing G, Mould DL, Clay M...   \n",
       "6   Doing G, Koeppen K, Occipinti P, Harty CE, Hog...   \n",
       "7   Lee AJ, Park Y, Doing G, Hogan DA, Greene CS. ...   \n",
       "8   Neff SL, Hampton TH, Puerner C, Cengher L, Doi...   \n",
       "9   Lee AJ, Reiter T, Doing G, Oh J, Hogan DA, Gre...   \n",
       "10  Lee AJ, Mould DL, Crawford J, Hu D, Powers RK,...   \n",
       "11  Lee AJ, Doing G, Neff SL, Reiter T, Hogan DA, ...   \n",
       "12  Doing G, Lee AJ, Neff SL, Reiter T, Holt JD, S...   \n",
       "13  Neff SL, Doing G, Reiter T, Hampton TH, Greene...   \n",
       "14  Taub L, Hampton TH, Sarkar S, Doing G, Neff SL...   \n",
       "15  Doing G, Shanbhag P, Bell I, Cassidy S, Motaki...   \n",
       "\n",
       "                                             url_slug  \\\n",
       "0   Unsupervised-Extraction-of-Stable-Expression-S...   \n",
       "1   Draft-Genome-Sequence-of-a-Violacein-Producing...   \n",
       "2   A-Multimodal-Strategy-Used-by-a-Large-c-di-GMP...   \n",
       "3   Genome-Sequences-of-Phenotypically-Distinct-Ja...   \n",
       "4   PathCORE-T:-identifying-and-visualizing-global...   \n",
       "5   Ethanol-Stimulates-Trehalose-Production-throug...   \n",
       "6   Conditional-antagonism-in-co-cultures-of-Pseud...   \n",
       "7   Correcting-for-experiment-specific-variability...   \n",
       "8   CF-Seq,-an-accessible-web-application-for-rapi...   \n",
       "9   Using-genome-wide-expression-compendia-to-stud...   \n",
       "10  SOPHIE:-Generative-Neural-Networks-Separate-Co...   \n",
       "11  Compendium-Wide-Analysis-of-Pseudomonas-aerugi...   \n",
       "12  Computationally-Efficient-Assembly-of-Pseudomo...   \n",
       "13  Pseudomonas-aeruginosa-transcriptome-analysis-...   \n",
       "14  E-PathDash-pathway-activation-analysis-of-publ...   \n",
       "15   TEAL-seq-Targeted-Expression-Analysis-Sequencing   \n",
       "\n",
       "                                            paper_url  \n",
       "0   http://academicpages.github.io/files/Unsupervi...  \n",
       "1   http://academicpages.github.io/files/Draft Gen...  \n",
       "2   http://academicpages.github.io/files/A Multimo...  \n",
       "3   http://academicpages.github.io/files/Draft Gen...  \n",
       "4   http://academicpages.github.io/PathCORE-T_ ide...  \n",
       "5   http://academicpages.github.io/Ethanol Stimula...  \n",
       "6   http://academicpages.github.io/Conditional ant...  \n",
       "7          http://academicpages.github.io/giaa117.pdf  \n",
       "8   http://academicpages.github.io/CF-Seq, an acce...  \n",
       "9   http://academicpages.github.io/Using genome-wi...  \n",
       "10  http://academicpages.github.io/SOPHIE_ Generat...  \n",
       "11  http://academicpages.github.io/Compendium-Wide...  \n",
       "12  http://academicpages.github.io/Computationally...  \n",
       "13  http://academicpages.github.io/neff_and_doing-...  \n",
       "14  http://academicpages.github.io/taub-et-al-2024...  \n",
       "15        http://academicpages.github.io/TEAL-seq.pdf  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "publications = pd.read_csv(\"publications.tsv\", sep=\"\\t\", header=0)\n",
    "publications\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Escape special characters\n",
    "\n",
    "YAML is very picky about how it takes a valid string, so we are replacing single and double quotes (and ampersands) with their HTML encoded equivilents. This makes them look not so readable in raw format, but they are parsed and rendered nicely."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "html_escape_table = {\n",
    "    \"&\": \"&amp;\",\n",
    "    '\"': \"&quot;\",\n",
    "    \"'\": \"&apos;\"\n",
    "    }\n",
    "\n",
    "def html_escape(text):\n",
    "    \"\"\"Produce entities within text.\"\"\"\n",
    "    return \"\".join(html_escape_table.get(c,c) for c in text)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating the markdown files\n",
    "\n",
    "This is where the heavy lifting is done. This loops through all the rows in the TSV dataframe, then starts to concatentate a big string (```md```) that contains the markdown for each type. It does the YAML metadata first, then does the description for the individual page."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2017-07-26\n",
      "2018-01-04\n",
      "2018-03-26\n",
      "2018-01-18\n",
      "2018-07-03\n",
      "2019-05-22\n",
      "2020-08-19\n",
      "2020-11-03\n",
      "2022-06-16\n",
      "2022-08-10\n",
      "2022-10-20\n",
      "2023-02-23\n",
      "2023-02-23\n",
      "2024-04-02\n",
      "2024-10-18\n",
      "2024-12-01\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "for row, item in publications.iterrows():\n",
    "    print(item.pub_date)\n",
    "    md_filename = str(item.pub_date) + \"-\" + item.url_slug + \".md\"\n",
    "    html_filename = str(item.pub_date) + \"-\" + item.url_slug\n",
    "    year = item.pub_date[:4]\n",
    "    \n",
    "    ## YAML variables\n",
    "    \n",
    "    md = \"---\\ntitle: \\\"\"   + item.title + '\"\\n'\n",
    "    \n",
    "    md += \"\"\"collection: publications\"\"\"\n",
    "    \n",
    "    md += \"\"\"\\npermalink: /publication/\"\"\" + html_filename\n",
    "    \n",
    "    if len(str(item.excerpt)) > 5:\n",
    "        md += \"\\nexcerpt: '\" + html_escape(item.excerpt) + \"'\"\n",
    "    \n",
    "    md += \"\\ndate: \" + str(item.pub_date) \n",
    "    \n",
    "    md += \"\\nvenue: '\" + html_escape(item.venue) + \"'\"\n",
    "    \n",
    "    if len(str(item.paper_url)) > 5:\n",
    "        md += \"\\npaperurl: '\" + item.paper_url + \"'\"\n",
    "    \n",
    "    md += \"\\ncitation: '\" + html_escape(item.citation) + \"'\"\n",
    "    \n",
    "    md += \"\\n---\"\n",
    "    \n",
    "    ## Markdown description for individual page\n",
    "        \n",
    "    if len(str(item.excerpt)) > 5:\n",
    "        md += \"\\n\" + html_escape(item.excerpt) + \"\\n\"\n",
    "    \n",
    "    if len(str(item.paper_url)) > 5:\n",
    "        md += \"\\n[Download paper here](\" + item.paper_url + \")\\n\" \n",
    "        \n",
    "    md += \"\\nRecommended citation: \" + item.citation\n",
    "    \n",
    "    md_filename = os.path.basename(md_filename)\n",
    "       \n",
    "    with open(\"../_publications/\" + md_filename, 'w') as f:\n",
    "        f.write(md)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These files are in the publications directory, one directory below where we're working from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18-Draft-Genome-Sequence-of-a-Violacein-Producing-Iodobacter-sp.-from-the-Hudson-Valley-Watershed.md\n",
      "18-PathCORE-T:-identifying-and-visualizing-globally-co-occurring-pathways-in-large-transcriptomic-compendia.md\n",
      "20-Correcting-for-experiment-specific-variability-in-expression-compendia-can-remove-underlying-signals.md\n",
      "2017-07-26-Unsupervised-Extraction-of-Stable-Expression-Signatures-from-Public-Compendia-with-an-Ensemble-of-Neural-Networks.md\n",
      "2017-Unsupervised-Extraction-of-Stable-Expression-Signatures-from-Public-Compendia-with-an-Ensemble-of-Neural-Networks.md\n",
      "2018-01-04-Draft-Genome-Sequence-of-a-Violacein-Producing-Iodobacter-sp.-from-the-Hudson-Valley-Watershed.md\n",
      "2018-01-18-Genome-Sequences-of-Phenotypically-Distinct-Janthinobacterium-sp.-Isolates-Cultured-from-the-Hudson-Valley-Watershed.md\n",
      "2018-03-26-A-Multimodal-Strategy-Used-by-a-Large-c-di-GMP-Network.md\n",
      "2018-07-03-PathCORE-T:-identifying-and-visualizing-globally-co-occurring-pathways-in-large-transcriptomic-compendia.md\n",
      "2018-A-Multimodal-Strategy-Used-by-a-Large-c-di-GMP-Network.md\n",
      "2018-Genome-Sequences-of-Phenotypically-Distinct-Janthinobacterium-sp.-Isolates-Cultured-from-the-Hudson-Valley-Watershed.md\n",
      "2019-05-22-Ethanol-Stimulates-Trehalose-Production-through-a-SpoT-DksA-AlgU-Dependent-Pathway-in-Pseudomonas-aeruginosa.md\n",
      "2019-Ethanol-Stimulates-Trehalose-Production-through-a-SpoT-DksA-AlgU-Dependent-Pathway-in-Pseudomonas-aeruginosa.md\n",
      "2020-08-19-Conditional-antagonism-in-co-cultures-of-Pseudomonas-aeruginosa-and-Candida-albicans:-An-intersection-of-ethanol-and-phosphate-signaling-distilled-from-dual-seq transcriptomics.md\n",
      "2020-11-03-Correcting-for-experiment-specific-variability-in-expression-compendia-can-remove-underlying-signals.md\n",
      "2020-Conditional-antagonism-in-co-cultures-of-Pseudomonas-aeruginosa-and-Candida-albicans:-An-intersection-of-ethanol-and-phosphate-signaling-distilled-from-dual-seq transcriptomics.md\n",
      "2022-06-16-CF-Seq,-an-accessible-web-application-for-rapid-re-analysis-of-cystic-fibrosis-pathogen-RNA-sequencing-studies.md\n",
      "2022-08-10-Using-genome-wide-expression-compendia-to-study-microorganisms.md\n",
      "2022-10-20-SOPHIE:-Generative-Neural-Networks-Separate-Common-and-Specific-Transcriptional-Responses. .md\n",
      "2022-CF-Seq,-an-accessible-web-application-for-rapid-re-analysis-of-cystic-fibrosis-pathogen-RNA-sequencing-studies.md\n",
      "2022-SOPHIE:-Generative-Neural-Networks-Separate-Common-and-Specific-Transcriptional-Responses. .md\n",
      "2023-02-23-Compendium-Wide-Analysis-of-Pseudomonas-aeruginosa-Core-and-Accessory-Genes-Reveals-Transcriptional-Patterns-across-Strains-PAO1-and-PA14.md\n",
      "2023-02-23-Computationally-Efficient-Assembly-of-Pseudomonas-aeruginosa-Gene-Expression-Compendia.md\n",
      "2023-Compendium-Wide-Analysis-of-Pseudomonas-aeruginosa-Core-and-Accessory-Genes-Reveals-Transcriptional-Patterns-across-Strains-PAO1-and-PA14.md\n",
      "2023-Computationally-Efficient-Assembly-of-Pseudomonas-aeruginosa-Gene-Expression-Compendia.md\n",
      "2024-04-02-Pseudomonas-aeruginosa-transcriptome-analysis-of-metal-restriction-in-ex-vivo-cystic-fibrosis-sputum.md\n",
      "2024-10-18-E-PathDash-pathway-activation-analysis-of-publicly-available-pathogen-gene-expression-data.md\n",
      "2024-12-01-TEAL-seq-Targeted-Expression-Analysis-Sequencing.md\n",
      "22-Using-genome-wide-expression-compendia-to-study-microorganisms.md\n"
     ]
    }
   ],
   "source": [
    "!ls ../_publications/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat: ../_publications/17-Unsupervised-Extraction-of-Stable-Expression-Signatures-from-Public-Compendia-with-an-Ensemble-of-Neural-Networks.md: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!cat ../_publications/17-Unsupervised-Extraction-of-Stable-Expression-Signatures-from-Public-Compendia-with-an-Ensemble-of-Neural-Networks.md"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
